README.md
# GScraper

* [github.com/postmodern/gscraper](http://github.com/postmodern/gscraper/)
* [github.com/postmodern/gscraper/issues](http://github.com/postmodern/gscraper/issues)
* Postmodern (postmodern.mod3 at gmail.com)

## Description
  
GScraper is a web-scraping interface to various Google Services.

## Features
  
* Supports the Google Search service.
  * Provides access to search results and ranks.
  * Provides access to the Sponsored Links.
* Provides HTTP access with custom User-Agent strings.
* Provides proxy settings for HTTP access.

## Requirements

* mechanize >= 0.9.0

## Install

    $ sudo gem install gscraper

## Examples

Basic query:

    q = GScraper::Search.query(:query => 'ruby')

Advanced query:

    q = GScraper::Search.query(:query => 'ruby') do |q|
      q.without_words = 'is'
      q.within_past_day = true
      q.numeric_range = 2..10
    end

Queries from URLs:

    q = GScraper::Search.query_from_url('http://www.google.com/search?as_q=ruby&as_epq=&as_oq=rails&as_ft=i&as_qdr=all&as_occt=body&as_rights=%28cc_publicdomain%7Ccc_attribute%7Ccc_sharealike%7Ccc_noncommercial%29.-%28cc_nonderived%29')

    q.query # => "ruby"
    q.with_words # => "rails"
    q.occurrs_within # => :title
    q.rights # => :cc_by_nc

Getting the search results:

    q.first_page.select do |result|
      result.title =~ /Blog/
    end

    q.page(2).map do |result|
      result.title.reverse
    end

    q.result_at(25) # => Result

    q.top_result # => Result

A Result object contains the rank, title, summary, cahced URL, similiar
query URL and link URL of the search result.

    page = q.page(2)

    page.urls # => [...]
    page.summaries # => [...]
    page.ranks_of { |result| result.url =~ /^https/ } # => [...]
    page.titles_of { |result| result.summary =~ /password/ } # => [...]
    page.cached_pages # => [...]
    page.similar_queries # => [...]

Iterating over the search results:

    q.each_on_page(2) do |result|
      puts result.title
    end

    page.each do |result|
      puts result.url
    end

Iterating over the data within the search results:

    page.each_title do |title|
      puts title
    end

    page.each_summary do |text|
      puts text
    end

Selecting search results:

    page.results_with do |result|
      ((result.rank > 2) && (result.rank < 10))
    end

    page.results_with_title(/Ruby/i) # => [...]

Selecting data within the search results:

    page.titles # => [...]

    page.summaries # => [...]

Selecting the data of search results based on the search result:

    page.urls_of do |result|
      result.description.length > 10
    end

Selecting the Sponsored Links of a Query:

    q.sponsored_links # => [...]

    q.top_sponsored_link # => SponsoredAd

Setting the User-Agent globally:

    GScraper.user_agent # => nil
    GScraper.user_agent = 'Awesome Browser v1.2'

## License

GScraper - A web-scraping interface to various Google Services.

Copyright (c) 2007-2009 Hal Brodigan (postmodern.mod3 at gmail.com)

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA



History.txt
=== 0.2.4 / 2009-03-18

* Fixed a bug in SponsoredAd#direct_url where direct links we're not
  being URI escaped.
* Removed last references to Hpricot, replacing them with Nokogiri.

=== 0.2.3 / 2009-01-27

* Fixed a bug in GScraper::Search::WebQuery#page, when the search query
  returned less results than the expected results-per-page.

=== 0.2.2 / 2009-01-14

* Updated GScraper::Search::WebQuery to use Nokogiri properly.

=== 0.2.1 / 2008-08-27

* Updated XPath queries in GScraper::Search::WebQuery for new Google (tm)
  Search Result HTML schema.

=== 0.2.0 / 2008-05-10

* Removed GScraper::WebAgent.
* Added GScraper::Page and GScraper::HasPages.
* GScraper::Search::Result#page and GScraper::Search::Result#cached_page
  no longer receives blocks.
* GScraper::SponsoredLinks.new and GScraper::Search::Page.new now take 
  blocks.
* Renamed GScraper::Search::Query to GScraper::Search::WebQuery.
* Added GScraper::Search::Query which supports building query expressions.
* GScraper::Search::WebQuery#page and GScraper::Search::WebQuery#sponsored_links
  no longer take blocks.
* Added GScraper::Search::AJAXQuery.
* Replaced Unit Tests with Rspec specifications.

=== 0.1.8 / 2008-04-30

* Added the GScraper.user_agent_alias=(name) method.
* Added URI::HTTP::QueryParams module.
* Changed license from MIT to GPL-2.

=== 0.1.7 / 2008-04-28

* Added support for specifing Search modifiers.

  Search.query(:filetype => :xls)

* Added the Search::Result#page method.

=== 0.1.6 / 2008-03-15

* Renamed GScraper.http_agent to GScraper.web_agent.
* Added GScraper.proxy for global proxy configuration.
* Added the WebAgent module.
* Renamed Search::Query#first_result to Search::Query#top_result.
* Updated Search::Query#page logic for the new DOM layout being used.
* Added support for Sponsored Ad scraping.
  * Added the methods Query#sponsored_links and Query#top_sponsored_link.
* Added examples to README.txt.

=== 0.1.5 / 2007-12-29

* Fixed class inheritance in gscraper/extensions/uri/http.rb, found by
  sanitybit.

=== 0.1.4 / 2007-12-23

* Added Search::Query#result_at for easier access of a single result at
  a given index.
* Adding scraping of the "Cached" and "Similar Pages" URLs of Search
  Results.
* Added methods to Search::Page for accessing cached URLs, cached pages,
  similar query URLs and similar Queries in mass.
* Search::Query#page and Search::Query#first_page now can receive blocks.
* Improved the formating of URL query parameters.
* Added more unit-tests.
* Fixed scraping of Search Result summaries.
* Fixed various bugs in Search::Query uncovered during unit-testing.
* Fixed typos in Search::Page's documentation.

=== 0.1.3 / 2007-12-22

* Added the Search::Page class, which contains many of convenance methods
  for searching through the results within a Page.

=== 0.1.2 / 2007-12-22

* Fixed a bug related to extracting the correct content-rights from search
  query URLs.
* Added GScraper.user_agent_aliases.

=== 0.1.1 / 2007-12-21

* Forgot to include lib/gscraper/version.rb.

=== 0.1.0 / 2007-12-20

* Initial release.
* Supports the Google Search service.



lib/gscraper/gscraper.rb
#
#--
# GScraper - A web-scraping interface to various Google Services.
#
# Copyright (c) 2007-2009 Hal Brodigan (postmodern.mod3 at gmail.com)
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
#++
#

require 'uri/http'
require 'mechanize'
require 'nokogiri'
require 'open-uri'

module GScraper
  # Common proxy port.
  COMMON_PROXY_PORT = 8080

  #
  # Returns the +Hash+ of proxy information.
  #
  def GScraper.proxy
    @@gscraper_proxy ||= {:host => nil, :port => COMMON_PROXY_PORT, :user => nil, :password => nil}
  end

  #
  # Creates a HTTP URI based from the given _proxy_info_ hash. The
  # _proxy_info_ hash defaults to Web.proxy, if not given.
  #
  # _proxy_info_ may contain the following keys:
  # <tt>:host</tt>:: The proxy host.
  # <tt>:port</tt>:: The proxy port. Defaults to COMMON_PROXY_PORT,
  #                  if not specified.
  # <tt>:user</tt>:: The user-name to login as.
  # <tt>:password</tt>:: The password to login with.
  #
  def GScraper.proxy_uri(proxy_info=GScraper.proxy)
    if GScraper.proxy[:host]
      return URI::HTTP.build(
        :host => GScraper.proxy[:host],
        :port => GScraper.proxy[:port],
        :userinfo => "#{GScraper.proxy[:user]}:#{GScraper.proxy[:password]}",
        :path => '/'
      )
    end
  end
  
  #
  # Returns the supported GScraper User-Agent Aliases.
  #
  def GScraper.user_agent_aliases
    WWW::Mechanize::AGENT_ALIASES
  end

  #
  # Returns the GScraper User-Agent
  #
  def GScraper.user_agent
    @@gscraper_user_agent ||= GScraper.user_agent_aliases['Windows IE 6']
  end

  #
  # Sets the GScraper User-Agent to the specified _agent_.
  #
  def GScraper.user_agent=(agent)
    @@gscraper_user_agent = agent
  end

  #
  # Sets the GScraper User-Agent using the specified user-agent alias
  # _name_.
  # 
  def GScraper.user_agent_alias=(name)
    @@gscraper_user_agent = GScraper.user_agent_aliases[name.to_s]
  end

  #
  # Opens the _uri_ with the given _options_. The contents of the _uri_
  # will be returned.
  #
  # _options_ may contain the following keys:
  # <tt>:user_agent_alias</tt>:: The User-Agent Alias to use.
  # <tt>:user_agent</tt>:: The User-Agent String to use.
  # <tt>:proxy</tt>:: A +Hash+ of proxy information which may
  #                   contain the following keys:
  #                   <tt>:host</tt>:: The proxy host.
  #                   <tt>:port</tt>:: The proxy port.
  #                   <tt>:user</tt>:: The user-name to login as.
  #                   <tt>:password</tt>:: The password to login with.
  #
  #   GScraper.open_uri('http://www.hackety.org/')
  #
  #   GScraper.open_uri('http://tenderlovemaking.com/',
  #     :user_agent_alias => 'Linux Mozilla')
  #   GScraper.open_uri('http://www.wired.com/',
  #     :user_agent => 'the future')
  #
  def GScraper.open_uri(uri,options={})
    headers = {}

    if options[:user_agent_alias]
      headers['User-Agent'] = WWW::Mechanize::AGENT_ALIASES[options[:user_agent_alias]]
    elsif options[:user_agent]
      headers['User-Agent'] = options[:user_agent]
    elsif GScraper.user_agent
      headers['User-Agent'] = GScraper.user_agent
    end

    proxy = (options[:proxy] || GScraper.proxy)
    if proxy[:host]
      headers[:proxy] = GScraper.proxy_uri(proxy)
    end

    return Kernel.open(uri,headers)
  end

  #
  # Similar to GScraper.open_uri but returns a Nokogiri::HTML document.
  #
  def GScraper.open_page(uri,options={})
    Nokogiri::HTML(GScraper.open_uri(uri,options))
  end

  #
  # Creates a new WWW::Mechanize agent with the given _options_.
  #
  # _options_ may contain the following keys:
  # <tt>:user_agent_alias</tt>:: The User-Agent Alias to use.
  # <tt>:user_agent</tt>:: The User-Agent string to use.
  # <tt>:proxy</tt>:: A +Hash+ of proxy information which may
  #                   contain the following keys:
  #                   <tt>:host</tt>:: The proxy host.
  #                   <tt>:port</tt>:: The proxy port.
  #                   <tt>:user</tt>:: The user-name to login as.
  #                   <tt>:password</tt>:: The password to login with.
  #
  #   GScraper.web_agent
  #
  #   GScraper.web_agent(:user_agent_alias => 'Linux Mozilla')
  #   GScraper.web_agent(:user_agent => 'Google Bot')
  #
  def GScraper.web_agent(options={},&block)
    agent = WWW::Mechanize.new

    if options[:user_agent_alias]
      agent.user_agent_alias = options[:user_agent_alias]
    elsif options[:user_agent]
      agent.user_agent = options[:user_agent]
    elsif GScraper.user_agent
      agent.user_agent = GScraper.user_agent
    end

    proxy = (options[:proxy] || GScraper.proxy)
    if proxy[:host]
      agent.set_proxy(proxy[:host],proxy[:port],proxy[:user],proxy[:password])
    end

    block.call(agent) if block
    return agent
  end
end
